client/app/bundles/course/level/components/LevelsTable.tsx

```diff
@@ -0,0 +1,158 @@
+import { FC } from 'react';
+import { defineMessages, FormattedMessage } from 'react-intl';
+import { Add, Delete, Undo } from '@mui/icons-material';
+import {
+  Button,
+  IconButton,
+  Table,
+  TableBody,
+  TableCell,
+  TableFooter,
+  TableHead,
+  TableRow,
+  TextField,
+  Tooltip,
+  Typography,
+} from '@mui/material';
+
+import useTranslation from 'lib/hooks/useTranslation';
+
+const translations = defineMessages({
+  levelHeader: {
+    id: 'course.level.Level.levelHeader',
+    defaultMessage: 'Level',
+  },
+  thresholdHeader: {
+    id: 'course.level.Level.thresholdHeader',
+    defaultMessage: 'Threshold',
+  },
+  addNewLevel: {
+    id: 'course.level.Level.addNewLevel',
+    defaultMessage: 'Add New Level',
+  },
+  orderedIncorrectly: {
+    id: 'course.level.Level.orderedIncorrectly',
+    defaultMessage:
+      'Levels will be sorted automatically when saved regardless of their order here.',
+  },
+});
+
+interface LevelsTableProps {
+  levels: { level: number; deleted: boolean }[];
```

Comment:
also, you might consider also adding the props of `draft` or `toBeAdded` over here.
The rationale is when you add some more levels, you cannot say for sure which level has been saved and which level is still to be saved. I think there should also be some kind of differentiation between to-be-saved and already-saved levels

Reply:
First point done.
Second point: toBeDeleted is just for filtering out what levels to send to backend to be saved (and to enable the undo feature). The saving of levels is done as 1 whole array of values, and none of them have to be specially marked to be deleted.

Reply:
Actually, if you see the implementation inside the BE on how they handle the saving, you'll notice that they actually delete those level objects in which thresholds are not included in the new array that you sent there. So no, they actually need to be marked as deleted in order for BE to process the deletion process correctly.

And, the second point I raised actually has nothing to do with the `toBeDeleted`. It's more towards the case when you add more levels, and those newly added levels will only become the official levels once you save it. I tried to add some levels, and then deleting it immediately, then I found that they will just be marked as deleted.

Imagine this scenario: how if I added 50 more levels, then decided to remove 45 of those new ones?
Here, I'm expecting those 45 to be vanished immediately, since it's not supposed to be there in the first place (not even recorded in our DB). But instead, I'll only get 45 rows of `toBeDeleted` entries, making the page inundated with unnecessary levels.

Reply:
Done, newly added levels will now disappear immediately.